### ISSUE TITLE
Returning Types smaller than Int32

### ISSUE THREAD

**Issue by mgaudet** [eclipse/omr#1901]
### Background 

In #1842 I first implemented a check that ensured all `ireturn` nodes had an `Int32` child type. However, this caused some failures in the compiler tests.

@Leonardo2718 pointed out that `returnOpCode` would return `ireturn` for all types smaller than or equal to Int32. 

After some discussion, and [an insightful comment from](https://github.com/eclipse/omr/pull/1842#issuecomment-339125394) @ymanton I opened #1888 and #1887, to track the creation of `sreturn` and `breturn` as opcodes.

#1842 was modified to allow Int16 and Int8 children to `ireturn`. 

I think I'd like to start some more discussion of this 

# The new thinking: 

On Friday, in pursuit of #1845 @rbanerjee and I were working on a test case for the conversions to Int8 and Int16 types. It was then that we realized that widening is an operation that needs to be sign bit aware: hence the existence of the `b2i` and `bu2i` / `s2i` and `su2i` pairs. The former of the pairs will sign-extend to Int32 and the latter will zero-extend. 

It's likely that `breturn` and `sreturn` would be widening operations, which means the possibility of having to create a `bureturn`and `sureturn`. We already have `iureturn` and `lureturn`, but I don't think that they are considered to be widenings (to be honest, I'm a little surprised to see them, unless `lureturn` is expected to widen an `Int32` child). 

In parallel, @rbanerjee has been investigating failures on s390x. It turns out that these failures are precisely missing widening operations and the s390x ireturn evaluator not correctly handling undersized arguments. The validator not raising the flag on these nodes caused more investigation was necessary. 

# The issues, summarized: 

The most major question is precisely: **What should the canonical way be to return integers smaller than Int32?**. 

I see three answers: 

* Mandate widening operation, and continue to use Int32 as the return type. 
* Mandate the CodeGenerators support widening non-Int32 children on `ireturn` 
* Implement `breturn`,`bureturn`,`sreturn`,`sureturn` 

There are attached subquestions: 

-  If we mandate a conversion to a wider type to be explicit in trees, what would the right return answer be for `OMR::ILOpCode::returnOpCode()`
-  How awkward is the multi-typed children of `ireturn` if we allow it do the widening in codegen -- is this a wart on the IL that isn't worth allowing?
- Can anyone fill in some of the history behind the unsigned returns? @vijaysun-omr / @mstoodle ? 
- @fjeremic Looking at ControlFlowEvaluator.cpp, it seems like the return evaluator tries to handle widening

    https://github.com/eclipse/omr/blob/7c3d3d72e6e741f8438f78604ee1eabe0aac9844/compiler/z/codegen/ControlFlowEvaluator.cpp#L1152-L1163

    But given we were able to fix our test cases by adding an explicit signed widening, is it possible that this is just a bug in the widening?

**Comment by ymanton** [comment]
> * Mandate widening operation, and continue to use Int32 as the return type.
> * Mandate the CodeGenerators support widening non-Int32 children on ireturn
> * Implement breturn,bureturn,sreturn,sureturn

(1) and (3) are effectively the same (the party creating the IL will either create `ireturns` fed with the correctly widened child, or will create the correct return type), except that (1) doesn't require any additional work in most cases, where as (3) is far more friendly to potential OMR Compiler users because it's not some arbitrary carry-over from the JVM bytecode spec that needs to be clearly specified/documented in OMR. (1) also has the downside that it happens to work for all of our current architectures but what happens on an arch that can only return 64-bit values and has to return them correctly signed/zeroed?

I don't see how (2) would work; how would the code generator know whether to sign- or zero-extend the child? It's easy for Java, signed is the only choice, but what if I have something like a `sadd` or a `bor` where the sign of the child is not specified by the operation at all? Since at the OMR layer the return type of a function lacks a sign it's also of no help; one way or another the party creating the IL needs to have the ability to tell the code generator the signedness of what it's returning.

**Comment by mstoodle** [comment]
Perhaps I'm oversimplifying, but shouldn't this just be a collaboration between the code generator (communicating whether various narrow types need to be widened for returns, which is presumably an "ABI thing") and the il generator (inserting appropriate unsigned/signed conversions according to the language's datatype requirements under the appropriate return opcode as dictated by the code generator) ?

**Comment by mgaudet** [comment]
That just seems... awkward to force every IlGenerator to have to be aware of all this negotiation required with a code-generator. 

To me, the IL itself should be the dominant interface between CodeGenerator and the hosting language runtime -- and if the IlGenerator wants to return a byte, it's up to the CodeGenerator to figure out ho.

**Comment by ymanton** [comment]
Asking the code generator how/what IL to create is another possibility sure, we do that in other places, but it means that low level details are leaking upward. If you allow that then you make things like persistent/serializable IL, cross compiling, multi-codegen, etc harder because you've already baked platform-specific details into your IL representation. Putting that aside I'd agree with @mgaudet, it's awkward, the IL should be expressive enough to communicate what's wanted to the code generator.

Personally, I'm leaning towards having signed/unsigned return variants for every size making the most sense.

**Comment by mstoodle** [comment]
The IL can be expressive enough; the question is what IL you generate.

All I was saying is that to answer that question depends both on the needs of the program you're converting to IL as well as the platform on which the program will run.

More explicitly, what I'm imagining is that the IL generator asks whether you can directly return some type (byte). If it can, the IL generator generates breturn with a byte typed operand. If it cannot, then the IL generator decides (based on the semantics of the program) whether it should generate an ireturn with an b2i of the return value, or an ireturn with some new bu2iu opcode on top of the return value.

You cannot, I don't think, avoid generating different IL for different platforms. So cross compiling IL to another platform is just not a requirement. It's a goal (we shouldn't arbitrarily decide to generate different IL for different platforms if it isn't necessary) but if it's necessary, then it's necessary.

**Comment by mstoodle** [comment]
How that gets implemented (whether even IL generator has to implement it) is different question. My basic statement is that you need information from both places to do the right thing. I don't believe you can fix it only in the IL generator or only in the code generator.

**Comment by ymanton** [comment]
> The IL can be expressive enough; the question is what IL you generate.
> All I was saying is that to answer that question depends both on the needs of the program you're converting to IL as well as the platform on which the program will run.

Doesn't that imply that the IL is not expressive enough though? If I can't generate the same IL for every platform and have things just work it means that the IL is not expressive enough on its own or even _portable_. I can't specify that I simply want to return a signed/unsigned X-bit integer and let the code generator figure out how/what it needs to make that happen, I have to let it dictate what IL I generate and be locked into that choice for the lifetime of the IL I generated. I know we're talking about returns here, but I think the same sort of issue exists for parameter passing. Likely the same sorts of issues will come up whenever we have to consider how to represent passing/returning aggregate value types, since that stuff is essentially dictated by the ABI.

This sort of thing deserves strong consideration because it's the sort of thing that will not be easily backed out of or corrected if we later find that it is not sufficient.

As far as this specific issue is concerned I think it's a consequence of our IL design choice, that we will ignore signedness wherever possible (most arithmetic and logical ops) and only care when necessary (comparisons)... that scheme is fine within a single method, but I think it breaks down when you have to pass things outside of the method you're compiling because then you can no longer follow your own conventions, you have to follow the conventions of the things you're calling and the things that called you. At that point you have to care about signedness, you won't always have the luxury of passing/returning exactly 8 or 16 or 32 bits, you'll have to pass/return what the platform dictates and you'll have to widen appropriately and you'll need to know the signedness, and currently the only way we can represent that in the IL is explicitly, which has all the downsides I mentioned above.

**Comment by Leonardo2718** [comment]
Personally, I tend to agree that this kind of bit-widening should be expressed in IL rather than be inferred by a code generator.

> More explicitly, what I'm imagining is that the IL generator asks whether you can directly return some type (byte). If it can, the IL generator generates breturn with a byte typed operand. If it cannot, then the IL generator decides (based on the semantics of the program) whether it should generate an ireturn with an b2i of the return value, or an ireturn with some new bu2iu opcode on top of the return value.

This brings up an interesting question about how much an IL generator should be platform aware. A similar question was raised in #566, regarding whether `aiadd` should be allowed on 64-bit platforms. The resolution was that it shouldn't be allowed. This necessarily introduces some level of platform awareness in IL generation.

Seeing this issue brought up again, I'm thinking we should have a more general discussion about how platform awareness should be handled in IL generation. Perhaps (and I'm only suggesting this to get the conversation started), we should define a new interface independent of the code generator and IL generator, specifically for answering these platform related questions.

This leaves the question of deciding how to _represent_ sub-integer return opcodes for this issue.

**Comment by mstoodle** [comment]
@ymanton :
> Doesn't that imply that the IL is not expressive enough though? If I can't generate the same IL for every platform and have things just work it means that the IL is not expressive enough on its own or even portable.

Portability has never been an unbreakable goal for our IL (evidence: is64Bit()). The goal of the IL is to express the operations the language wants to perform in a way the optimizer can reason about and in a way that the code generator can faithfully convert to native instructions to preserve the language's semantics. In the Java world, Java bytecodes are portable, but portability as an absolute requirement stops there (imo). There are also lots of runtime options (gc policy, compressed refs or not, etc.) that affect how we generate IL, so expecting a single canonical IL for every method will never be practical.

@ymanton :
> you have to follow the conventions of the things you're calling and the things that called you

Right, and those conventions may be different on different platforms and different for different languages, as you pointed out. Requiring a single canonical piece of IL to express those conventions for every possible language and every possible platform is unreasonable (imo). It will either force some languages to jump through hoops to convert into that pristine IL (probably inheriting semantics they don't want or don't need), or it will force some platform code generators to jump through hoops to operate "just like" the other platforms (without the optimizer's help to improve the efficiency of the hoops).

@ymanton :
> currently the only way we can represent that in the IL is explicitly, which has all the downsides I mentioned above.

And also the **up side** that it exposes any required sign extension operations to the optimizer. That's a very good reason, for example, to not include "write barriers" as an implied part of every indirect store to avoid introducing the iwrtbar opcode only when certain GC policies are enabled.

@Leonardo2718 :
> This brings up an interesting question about how much an IL generator should be platform aware.

Maybe, but in the way you expressed it, I consider this to be an implementation question.

If one accepts that IL is platform specific (which it is in every multi-language, multi-platform compiler I know of which, as you also pointed out, includes ours :) ) then the process of generating IL necessarily needs to have some mechanism to do the right things for the target platform. Whether those need to be implemented by every IL generator (true today) or could be abstracted behind (IL generator or some other) services like "genReturn(node, true /* isUnsigned */)" that query the code generator to build the right IL is "just" an implementation choice.

OTOH, if one only believes in a platform-neutral IL, then the answer is simple "not at all aware" :) .

**Comment by mgaudet** [comment]
I think this discussion wandered a bit further afield than I intended. Much of the above raises good points, and I am definitely in agreement that there are some things that must be negotiated during IlGeneration. 

But I'm really interested in how we want to have someone return a byte or a short. 

Personally: I'm heavily biased against having the IlGenerator have to query the code generator to figure out how to return a byte or short, mostly because I think it's a sufficiently simple situation that we should be able to have defined semantics. But I'd be remiss if I didn't mention my other concern: The ways in which this will complicate writing Tril tests. 

Right now Tril test are strings that are parsed and approximately map 1-1 to a set of trees as far as possible: How do I query the requirements for returning bytes in this world? I suppose the Tril IlGenerator could silently do the widening conversion, but philosophically it violates the goal of having as much control over the trees as possible. 

I don't think we'll solve the questions about IlGeneration here -- and certainly this has provided much food for thought about future work in Tril, as we'll need a way to capture dimensions like CopressedRefs where the IlGenerator needs to make different choices because of the mode.

I'd like to propose closing this off by suggesting that the right way to answer my original question is to implement `breturn`, `bureturn`, `sreturn` and `sureturn`.

**Comment by ymanton** [comment]
> The goal of the IL is to express the operations the language wants to perform in a way the optimizer can reason about and in a way that the code generator can faithfully convert to native instructions to preserve the language's semantics.

But in this case that's clearly not possible. Let's set aside portability for a sec and focus on expressiveness. The language wants to return 8 or 16-bit integers, the code generator(s) need to know the signedness of what's being returned to make it work, the language has that information (it knows the signedness of its types) but we're not allowing it to pass that to the CG via the IL, instead we're mandating that they establish a side-channel of communication so that the language can tell it what it wants and the code generator can tell it what and how to generate it. In the end the same bit of information found its way to the code generator, at the cost of having every IL generator go through this side channel and in the process tie its IL to a particular platform. We lose generality, make IL generation more complicated, and the gain is that the optimizer can see conversion, but if that's so important that it's worth doing it this way there are far better ways to accomplish that, which I'll elaborate on in a sec.

> Right, and those conventions may be different on different platforms and different for different languages, as you pointed out. Requiring a single canonical piece of IL to express those conventions for every possible language and every possible platform is unreasonable (imo). It will either force some languages to jump through hoops to convert into that pristine IL (probably inheriting semantics they don't want or don't need), or it will force some platform code generators to jump through hoops to operate "just like" the other platforms (without the optimizer's help to improve the efficiency of the hoops).

Not sure I follow, the conventions might be different, but the IL doesn't need to express the conventions at all, it expresses the intent (I want to return a signed/unsigned X-bit integer), the ABI conventions required to realize that are wholly contained in the code generator and it should have enough information to actually make that decision. I'd argue what we have now is a better example of "expressing conventions" in the IL, we're making the ABI conventions visible in the IL from the get-go.

> And also the up side that it exposes any required sign extension operations to the optimizer. That's a very good reason, for example, to not include "write barriers" as an implied part of every indirect store to avoid introducing the iwrtbar opcode only when certain GC policies are enabled.

Sure, we want an IL that isn't so high level it brings along with it unnecessary baggage, but at the same time you want an IL that isn't so low level that it requires the IL generator to immediately be aware of and have to cater to the code generator. If we want to expose more things to the optimizer that would better be accomplished by a proper lowering pass that turns platform neutral IL into more explicit IL later in the pipeline, after which it can run whatever opts make sense, rather than requiring the IL generator to have to generate it up front. I don't see what we're gaining by forcing IL generators to participate in this sort of lowering when we can simply allow the IL generator to better express itself so we can deal with the plumbing. That may not always be possible, but I don't think we should go down this path without actually considering how it affects more complicated things like aggregates. It's better to argue about it now than to have to correct it later.

**Comment by mstoodle** [comment]
I think my position and suggested solution may have been misunderstood, so let me be a bit more explicit. Even if not, at least it gets back to @mgaudet's optimistic request to discuss the actual problem at hand :) .

I'm suggesting that the IL generation process ask the code generator a question like: what is the opcode to return a value of "this" type for this platform. Based on that answer, the IL generation process either generates (if the opcode has the same basic data type as "this" type). In the case of a value of TR::Byte :
```
breturn
    <the byte to return>
```

or (if the language says I'm returning a signed byte) :
```
ireturn
   b2i
      <the byte to return>
```

or (if the language says I'm returning an unsigned byte) : 
```
ireturn
   bu2iu
      <the byte to return>
```

There is no "side channel" (at least from the angle from which I'm squinting). The generated IL represents exactly what needs to happen and the code generator gets a value it knows exactly how to deal with (either a byte or an already appropriately widened integer).

Through this collaboration, the IL generating process and the code generator can cause different IL to be generated depending on the platform and how the underlying language has used this return type.

Note that I used "IL generating process" to avoid implying that I would force every IL generator implementation to implement the logic to figure this out. That code can probably be written once and exposed as a service to IL generators everywhere to get the proper effect. With an appropriate API, the code generator could even do the "heavy" lifting under the direction of an IL generator.

The fragility Matthew is worried about is one of the reasons we haven't (after 19 years and hundreds of developer years of effort) already exposed TR IL as a persistent format for writing tests and the like. Similarly, it's one of the reasons I started tackling higher level APIs like JitBuilder that attempt to abstract some of the details away (amazingly I can hear the eyeballs rolling as I write that) by codifying some of the "IL generating process". But even at the JitBuilder level you run into problems. What code do you generate for something like Transaction() on a platform that does not support hardware transactional memory? Or for AtomicAdd() if the platform doesn't have one?

In the end, either you make the JIT after IL generation entirely responsible for platform knowledge, or you don't. Neither choice is without serious consequences for some scenarios. I would say we're currently rather invested in the "you don't" and it would not be a small amount of work to move all the way to the "you do" side. Just to be clear: I'm not saying we shouldn't aspire to generate IL as similarly across platforms in as many cases as possible. It's just not an unbreakable rule.

Having written all that out, I have the feeling that I may still be misunderstanding the problem, or the root concern, or that we're vehemently disagreeing on how we say very similar things. I'm still not quite sure which it is but I'm suspicious it isn't the last one no matter how much I'm hoping it is.

**Comment by mstoodle** [comment]
If a 64-bit platform, then presumably those (widening) trees turn to:

```
lreturn
   b2l
      <the byte to return>
```

and:

```
lreturn
   bu2lu
      <the byte to return>
```

**Comment by mgaudet** [comment]
(Just a side note: `breturn` doesn't currently exist)

**Comment by ymanton** [comment]
> I'm suggesting that the IL generation process ask the code generator a question like: what is the opcode to return a value of "this" type for this platform. Based on that answer, the IL generation process either generates (if the opcode has the same basic data type as "this" type). In the case of a value of TR::Byte
...
There is no "side channel" (at least from the angle from which I'm squinting). The generated IL represents exactly what needs to happen and the code generator gets a value it knows exactly how to deal with (either a byte or an already appropriately widened integer).

That's the side channel itself, the fact that the IL generator is trying to express its intent within the confines of the IL but can't and has to go to the code generator for something as low-level as how to pass/return primitives.

> But even at the JitBuilder level you run into problems. What code do you generate for something like Transaction() on a platform that does not support hardware transactional memory? Or for AtomicAdd() if the platform doesn't have one?

I think that's actually an example of a far better level of abstraction. Sure, not all platforms have TM or atomics, but the query as far as the IL generator is simple, do you support X and can I use X-IL? If the answer is yes the IL the generator emits is at the very least portable, it's the same IL across all platforms, the generator doesn't have to care that on this arch atomics are done via CAS and on that platform atomics are actually LL/SC. Likewise TM, the exact same IL works on all platforms because (at the moment anyway), it's expressive enough to allow IL generator to say what it wants/needs and the code generator(s) can turn it into a HW transaction.

> Having written all that out, I have the feeling that I may still be misunderstanding the problem, or the root concern, or that we're vehemently disagreeing on how we say very similar things. I'm still not quite sure which it is but I'm suspicious it isn't the last one no matter how much I'm hoping it is.

The concern is that this becomes far more distasteful once you get into more complicated scenarios. Try taking a crack at your earlier example with a more complicated "type":

The IL generation process asks the code generator a question like: what is the opcode to return a value of "this" type for this platform, where "this" type is a struct that looks like "{ Int8, Int32, Float, Int64, Double }" or "{ Int32 repeated 20 times }" or whatever you can think of. Suddenly the ABI requirements that we're exposing to IL generation are no longer so simple and require things passing and returning things through memory, temporary storage, implicit parameters, and so on. Even if you provide a convenience service to hide whatever is required to get the IL generated anyone who then wants to walk the IL they generated, for example as some sort of language specific opt pass, gets a mouth full of ABI details.

**Comment by fjeremic** [comment]
> @fjeremic Looking at ControlFlowEvaluator.cpp, it seems like the return evaluator tries to handle widening

@mgaudet looks like it only handles 32-bit to 64-bit widening based on the linkage used for the method being compiled. Seems it probably ties to abide by the ABI.

**Comment by mstoodle** [comment]
> Try taking a crack at your earlier example with a more complicated "type"

That's a better example, I think you're right, as it might drive to a different (better, more general) API. But I'm not 100% sure it will change my mind. I'm also not 100% sure my suggested approach wouldn't work since, if the compiler has been upgraded to work with aggregate values with fully fledged types, then maybe asking the code generator for the appropriate return opcode to return a value with an "aggregate" type really is the proper thing to do (because maybe there's an "aggregate return" opcode).

Note that I didn't recommend "always expose every detail of the ABI to the IL". I just suggested to "authorize" the IL generator to collaborate with the code generator (as representative for "the platform") in the same way it collaborates with the language. I would call that a more flexible class of designs.

On the other hand, your position (if I may put words in your mouth, sorry if they don't quite fit :) ) is that no ABI detail is desirable outside the code generator / platform-specific classes. My feeling is that this position is the more extreme one that has the potential to induce other (more painful?) trade-offs to be made in future.

Are you also intending to hide the ABI details from the optimizer when we inline? That also sounds like a potentially thorny position to take...

**Comment by ymanton** [comment]
> On the other hand, your position (if I may put words in your mouth, sorry if they don't quite fit :) ) is that no ABI detail is desirable outside the code generator / platform-specific classes. My feeling is that this position is the more extreme one that has the potential to induce other (more painful?) trade-offs to be made in future.
Are you also intending to hide the ABI details from the optimizer when we inline? That also sounds like a potentially thorny position to take...

My position is that we should progressively lower the IL where it makes sense to do so. I'm all for exposing things to the optimizer actually, but it should be done at a point in the pipeline where it minimizes complexity and repetition and maximizes opportunity. On the one hand we've pushed complicated opcodes like `new` and `monent` all the way to the code generators when there's a good argument to be made that they should be lowered earlier to not only expose opportunities to the optimizer but also to save us from having to implement basic evaluation for each and every code generator in one of the most tedious and error-prone environments possible. Likewise, I think exposing the ABI details at the front of the pipeline is equally troublesome, because now the complexity is exposed earlier than it needs to be to IL generators probably don't care to know the details. The fact that we're only dealing with simple primitives at the moment gives us a false sense of security that the complexity isn't too bad, but I don't want to end up in a situation where we're inflicting the same sort of pain on ourselves at both ends of the pipeline because we didn't choose the right representation when we had the chance.

Having nodes like `new` and `monent` makes things like Escape Analysis and lock optimizations easier, that's a good level of granularity for those opts, but that doesn't mean those nodes need to stay like that forever because the operations they represent hide a lot of code that might well be optimized if we exposed them, but more importantly it would significantly simplify evaluation. Similarly, does the Inliner need to see the ABI details of passing and returning structs by value? Maybe not, maybe it's easier to inline things w.r.t. value types when you can simply treat aggregates like their primitive counterparts instead of an explosion loads/stores/call parameters that's different on every platform, but again that doesn't mean that exposing that representation later wouldn't be beneficial so you can run opts that can actually do something meaningful, and I'm all for that.

**Comment by mstoodle** [comment]
In the interest of moving this forward, I can live with signed/unsigned version of return opcodes.

I'm not really convinced this solution won't have just as unfortunate potential implications on the future design of aggregate types (e.g. i really hope we won't feel it necessary to cook up new return opcodes to convey the signedness for each field of an aggregate), but I guess I'm willing to cross that bridge when we decide to resolve the 800 other issues we'll need to account for when we add aggregate types.

**Comment by ymanton** [comment]
> In the interest of moving this forward, I can live with signed/unsigned version of return opcodes.

@mgaudet,

As I said earlier I think the same problem exists with parameters, if we're passing integers and the ABI requires that they be widened we need to know how to do that. If we go with signed/unsigned return ops we would need a different scheme for parameters (maybe some sort of op that only conveys the _signedness_ of its child?).

Alternatively, we can communicate signedness in the method's return info and parameters, something like:

`signed/unsigned Int8 someMethod(signed/unsigned Int16, signed/unsigned Int32, ...)`

That way you can keep signless returns but you can still figure out how to widen them if necessary; likewise for the parameters. That means that our datatypes can stay signless, the returns can stay signless, but at least the signedness info is available if it's needed and it addresses the problem for parameters as well without too much disruption.

**Comment by mgaudet** [comment]
That is an excellent point not considered (or maybe I missed it in the earlier discussion 😓). I wonder if this issue is also coming out in #1943. 

I kind of like your idea @ymanton, of attaching this info to the 'signature'; though, I'm not familiar enough with how the signature/linkage code all links together to give a statement as to whether or not this would be easy or hard, or miss something fundamental.

**Comment by ymanton** [comment]
If you look here it seems like we have something like this already, where you can supposedly get a method's return type and it's signedness, amongst other things: https://github.com/eclipse/omr/blob/968837fbc8c8bc6e85020257a4e0c92f0a429136/compiler/compile/OMRMethod.hpp#L113-L130

We also have an iterator that you can use to go through a method's parameters; it returns the datatype of the parameter so that might be a good place to also return its signedness:
https://github.com/eclipse/omr/blob/968837fbc8c8bc6e85020257a4e0c92f0a429136/compiler/compile/OMRMethod.hpp#L96-L106

The code generator would then have to get its hands on this info wherever its required, like when evaluating a `breturn` or an `sreturn` or evaluating calls or accessing its own incoming parameters.

> though, I'm not familiar enough with how the signature/linkage code all links together to give a statement as to whether or not this would be easy or hard, or miss something fundamental.

I can foresee that we'd need something else for aggregates; it doesn't make sense to specify the signedness of an aggregate, you have to specify the signedness of the elements themselves. Some ABIs will want you to pass them through memory, others will allow you to pack them in registers, but some will want the individual elements each in their own registers, at which point you need to know how to widen each element. I think we would have to maintain the signedness info of the elements together with their datatypes in the definition of the aggregate type itself, which doesn't seem so bad actually.
